### PR TITLE
feat: Add `Range.__reversed__ ` and panic on range step=0

### PR DESCRIPTION
**Pull request by thewill-i-am**
Add a `reverse_in_place` method to the `Range` struct that reverses the iteration direction in place by negating `_step` and recomputing `_next` and `_stop` so the reversed range yields the same elements in opposite order. Mirrors the existing `array.reverse_in_place` pattern.

Fix a hang where iterating `range(a, b, 0)` would loop forever: the `_next += 0` cursor never advanced. The public `_range3` constructor now panics with "range() arg 3 must not be zero", and `reverse_in_place` panics on a zero-step range as a defensive guard. `panic` is imported deferred from `guppylang.std.platform` to avoid a circular import since `iter` loads early via `builtins`/`array`.

Tests cover 9 reverse cases (positive/negative/uneven/empty ranges) checked against `reversed(builtins.range(...))`, HUGR validation, double-reverse idempotency, and runtime panics for both `step=0` paths.

### CODE REVIEW

File: tests/integration/test_range.py [hunk 1]
```diff
@@ -78,3 +81,110 @@ def main() -> None:
...
+def test_range_reverse(run_int_fn):
+    """Reversing a range iterates the same elements as `reversed(range(...))`."""
+
+    @guppy
+    def main(start: int, stop: int, step: int) -> int:
+        total = 0
+        r = range(start, stop, step)
+        r.reverse_in_place()
+        for x in r:
+            total += x + 100  # Make the initial 0 obvious
```
Comment (maximilianruesch) [inline tests/integration/test_range.py]: (not a full review) How does this tell you in what order your elements are present? As I understand them, this test asserts that all elements are iterated over, but not what order they are in.

Comment (thewill-i-am) [reply tests/integration/test_range.py]: Hi @maximilianruesch 

I added this test because `reverse_in_place` mutates the internal `Range` state and needs
runtime coverage across edge cases like negative steps, uneven steps, empty ranges, and
off-by-one boundaries. The goal was to compare against Python's
`reversed(builtins.range(...))`.

You're right that the current sum-based check does not verify order.

Comment (thewill-i-am) [reply tests/integration/test_range.py]: @maximilianruesch 

I updated the test to use an order-sensitive accumulator instead, so sequences like
`3, 1` and `1, 3` now produce different results. 


The test is now order-sensitive because the accumulator is no longer a simple sum.

Previously, the test used:
```python
total += x + 100
That only verifies that the same values are produced, because addition is commutative:
1 + 3 == 3 + 1
So the test could pass even if reverse_in_place produced the right elements in the wrong order.
Now the test uses an order-sensitive fold:
total = total * 100 + x + 50
This makes the accumulated result depend on the exact sequence of values.
For example, for range(1, 5, 2), the original sequence is:
1, 3
The reversed sequence should be:
3, 1
With the new accumulator, the correct reversed sequence gives:
# 3, 1
total = 0
total = 0 * 100 + 3 + 50   # 53
total = 53 * 100 + 1 + 50  # 5351
But the wrong order gives a different result:
# 1, 3
total = 0
total = 0 * 100 + 1 + 50   # 51
total = 51 * 100 + 3 + 50  # 5153
So 5351 != 5153, and the test would fail if the range were not actually reversed.
The expected value uses the same encoding over Python's reference behavior:
return encode(reversed(builtins.range(a, b, s)))
```

So the test compares Guppy's reverse_in_place result against the order produced by
Python's reversed(range(...)).

File: guppylang/src/guppylang/std/iter.py [hunk 2]
```diff
@@ -73,6 +73,35 @@ def __next__(self: Self @ owned) -> Option[tuple[int, Self]]:
         self._next += self._step
         return some((actual_next, self))
 
+    @guppy
+    @no_type_check
+    def reverse_in_place(self: Self) -> None:
```
Comment (CalMacCQ) [inline guppylang/src/guppylang/std/iter.py]: In Python, `list.reverse()` also acts in place. Therefore maybe we don't need to call it `reverse_in_place`. What do you think @nicolaassolini-qntm ?

File: guppylang/src/guppylang/std/iter.py [hunk 3]
```diff
@@ -111,4 +142,12 @@ def range(start: int, stop: int = 0, step: int = 1) -> Range:
     ``start`` defaults to ``0`` and ``step`` defaults to ``1``. If the provided ``stop``
     value is comptime known, then the returned iterator will have a static size
     annotation and may for example be used inside array comprehensions.
+
+    Iterating with a ``step`` of ``0`` raises a runtime panic. The
+    ``Range.reverse_in_place`` method reverses the iteration direction in place.
```
Comment (nicolaassolini-qntm) [inline guppylang/src/guppylang/std/iter.py]:
```suggestion
    Iterating with a ``step`` of ``0`` raises a runtime panic. 
    
    Use the ``Range.reverse_in_place`` method to reverse the iteration direction in place.
```

File: guppylang/src/guppylang/std/iter.py [hunk 4]
```diff
@@ -73,6 +73,35 @@ def __next__(self: Self @ owned) -> Option[tuple[int, Self]]:
...
+            if diff > 0:
+                neg = -self._step
+                count = (diff + neg - 1) // neg
+                last = self._next + (count - 1) * self._step
+                self._stop = self._next - self._step
+            else:
+                self._stop = self._next
+                last = self._next
+            self._next = last
+            self._step = -self._step
```
Comment (nicolaassolini-qntm) [inline guppylang/src/guppylang/std/iter.py]: Just a refactor to make the function easier to follow (IMO)

I hope indentation is fine :)

```suggestion
    def reverse_in_place(self: Self) -> None:
        if self._step == 0:
            panic("range.reverse_in_place: step is zero")

        diff = self._stop - self._next

        # The range is empty when diff and step have different signs,
        # or when the difference is zero.
        if diff == 0 or (diff < 0) != (self._step < 0):
            last = self._next
            self._stop = self._next
        else:
            step = self._step
            # Change diff and step sign to ensure that count is correct
            if diff < 0:
                diff = -diff
                step = -step

            count = (diff + step - 1) // step
            last = self._next + (count - 1) * self._step
            self._stop = self._next - self._step

        self._next = last
        self._step = -self._step
 ```

File: tests/integration/test_range.py [hunk 5]
```diff
@@ -78,3 +81,122 @@ def main() -> None:
...
+        r.reverse_in_place()
+        for x in r:
+            total = total * 100 + x + 50
+        return total
+
+    def encode(xs) -> int:
+        total = 0
+        for x in xs:
+            total = total * 100 + x + 50
+        return total
```
Comment (nicolaassolini-qntm) [inline tests/integration/test_range.py:103]: Maybe a brief comment to explain the logic here?

File: tests/integration/test_range.py [hunk 6]
```diff
@@ -78,3 +81,122 @@ def main() -> None:
...
+    @guppy
+    def main(start: int, stop: int, step: int) -> int:
+        total = 0
+        r = range(start, stop, step)
+        r.reverse_in_place()
+        for x in r:
+            total += x
+        return total
+
+    validate(main.compile_function())
```
Comment (nicolaassolini-qntm) [inline tests/integration/test_range.py]: I think this is redundant after the change I suggested above (or below chronologically on GitHub)

File: tests/integration/test_range.py [hunk 7]
```diff
@@ -78,3 +81,122 @@ def main() -> None:
...
+
+    def encode(xs) -> int:
+        total = 0
+        for x in xs:
+            total = total * 100 + x + 50
+        return total
+
+    def expected(a: int, b: int, s: int) -> int:
+        return encode(reversed(builtins.range(a, b, s)))
+
```
Comment (nicolaassolini-qntm) [inline tests/integration/test_range.py:107]:
```suggestion

validate(main.compile_function())
```

**Review by nicolaassolini-qntm — CHANGES_REQUESTED** [review]
Thanks!
I think the implementation is good. Here some suggestions

**Comment by thewill-i-am** [comment]
@nicolaassolini-qntm @CalMacCQ , Thank you for your time spent on the review; I'll implement the suggestions. I think they should be ready by Monday.

File: tests/integration/test_range.py (hunk 6 above)
Comment (thewill-i-am) [reply tests/integration/test_range.py]: Yes, agreed. Since `test_range_reverse` already compiles and emulates the reversed range
across multiple cases, the validate-only test is redundant now. I removed the whole
`test_range_reverse_validate` test rather than only the `validate(main.compile_function())`

File: guppylang/src/guppylang/std/iter.py (hunk 4 above)
Comment (thewill-i-am) [reply guppylang/src/guppylang/std/iter.py]: @nicolaassolini-qntm. 

I tried the suggested refactor locally, but it seems to change the behavior for negative-step ranges.

For example, `range(5, 0, -1)` yields `5, 4, 3, 2, 1`, so after `reverse_in_place()` I would expect it to yield `1, 2, 3, 4, 5`, matching `reversed(builtins.range(5, 0, -1))`.

With the suggested refactor, that case becomes empty instead. I think the issue comes from using one shared `diff = stop - next` formula for both positive and negative steps, which makes the signed division/modulo cases tricky.

Could you confirm that the expected behavior for `range(5, 0, -1).reverse_in_place()` is `1, 2, 3, 4, 5`?

Comment (thewill-i-am) [reply guppylang/src/guppylang/std/iter.py]: @nicolaassolini-qntm More explanation here: 

For `range(5, 0, -1).reverse_in_place()`:

- The current/original implementation produces: `1, 2, 3, 4, 5`
- The suggested refactor produces: an empty range

For `range(1, 5, 2)` reversed twice:

- The current/original implementation produces: `1, 3`
- The suggested refactor produces: an empty range

Could you confirm which behavior you expect? My understanding is that `reverse_in_place`
should match `reversed(builtins.range(...))`, and reversing twice should restore the
original sequence.

File: guppylang/src/guppylang/std/iter.py (hunk 2 above)
Comment (nicolaassolini-qntm) [reply guppylang/src/guppylang/std/iter.py]: I agree.

I also think that having range.reserve() is compatible with a future reversed(i: Iterable) implemented through protocols. (@mark-koch  ?)

File: guppylang/src/guppylang/std/iter.py (hunk 4 above)
Comment (nicolaassolini-qntm) [reply guppylang/src/guppylang/std/iter.py]: My bad, I updated the suggestion, now it should be fine (the issue was `count = (diff + step - 1) // step` with negative `diff` and `step`

Comment (thewill-i-am) [reply guppylang/src/guppylang/std/iter.py]: it's working now @nicolaassolini-qntm !!!

I removed a redundant `count` assignment. The first `count = ...` was immediately
overwritten by the next line, so it had no effect, and it could be misleading for negative
steps because `step` could still be negative there. The remaining formula uses
`distance = abs(diff)` and `step = abs(self._step)`, which is the intended calculation.

File: guppylang/src/guppylang/std/iter.py (hunk 2 above)
Comment (thewill-i-am) [reply guppylang/src/guppylang/std/iter.py]: That makes sense; I'm happy to rename the method, but I think that's a longer discussion. I'll wait for you to decide. I'll be waiting for your decision. @nicolaassolini-qntm @CalMacCQ @mark-koch

File: tests/integration/test_range.py (hunk 5 above)
Comment (thewill-i-am) [reply tests/integration/test_range.py:103]: `# Use an order-sensitive fold to distinguish e.g. [3, 1] from [1, 3].` Done

File: tests/integration/test_range.py (hunk 7 above)
Comment (thewill-i-am) [reply tests/integration/test_range.py:107]: I think this suggestion is outdated / misplaced now. This line is inside the
`expected(...)` helper, so replacing it with `validate(main.compile_function())` would
break the test because `expected` would no longer return the Python reference value.

The validate-only test was removed separately because `test_range_reverse` already compiles
and emulates the function through `run_int_fn`.  Please correct me if I am wrong

Comment (nicolaassolini-qntm) [reply tests/integration/test_range.py:107]: I'm suggesting to add at line 107 `validate(main.compile_function())` before the list of `run_int_fn` just to see if a potential failing test is due to validation or due to an emulator fail

Comment (thewill-i-am) [reply tests/integration/test_range.py:107]: Oh, makes sense. Done !
It's fixed now !  -> Line 108

**Review by nicolaassolini-qntm — COMMENTED** [review]
Is good for me now; we'll let you know about the function name. Thanks again!

**Review by nicolaassolini-qntm — APPROVED** [review]

File: guppylang/src/guppylang/std/iter.py (hunk 2 above)
Comment (nicolaassolini-qntm) [reply guppylang/src/guppylang/std/iter.py]: Hi @thewill-i-am, We discussed and the proper name should be `__reversed__`. Thanks for the help!

Also, I edited your PR since this PR is not closing completely https://github.com/Quantinuum/guppylang/issues/2086 (your contribution is a good help for us in implementing that feature 😄 )

Comment (thewill-i-am) [reply guppylang/src/guppylang/std/iter.py]: @nicolaassolini-qntm Awesome! It was very interesting working on this, let me know if there's anything else I can help with.

**Comment by nicolaassolini-qntm** [comment]
@thewill-i-am I renamed the method and requested a Copilot review for a final check. If everything is fine, as I suppose, I will merge your contribution.

Thanks again. If you are willing to contribute again to our projects (guppy, but also tket2 and hugr), feel free to have a look at our open issues. We also use the label `P-help-wanted` on the issues that are more suitable for external contributors.

**Comment by thewill-i-am** [comment]
> @thewill-i-am I renamed the method and requested a Copilot review for a final check. If everything is fine, as I suppose, I will merge your contribution.
> 
> Thanks again. If you are willing to contribute again to our projects (guppy, but also tket2 and hugr), feel free to have a look at our open issues. We also use the label `P-help-wanted` on the issues that are more suitable for external contributors.

@nicolaassolini-qntm  Thanks! My plan is to keep contributing because I’m a big fan of the work you did with Guppy. :)

By the way, some tests are failing after the rename. Would you like me to take a look at that?
